graphql_compiler/schema_transformation/utils.py

```diff
@@ -39,16 +40,41 @@ class InvalidTypeNameError(SchemaTransformError):
     """
 
 
-class SchemaNameConflictError(SchemaTransformError):
-    """Raised when renaming or merging types or fields cause name conflicts.
+class SchemaMergeNameConflictError(SchemaTransformError):
+    """Raised when merging types or fields cause name conflicts.
 
-    This may be raised if a field or type is renamed to conflict with another field or type,
-    if two merged schemas share an identically named field or type, or if a
+    This may be raised if two merged schemas share an identically named field or type, or if a
     CrossSchemaEdgeDescriptor provided when merging schemas has an edge name that causes a
     name conflict with an existing field.
     """
 
 
+class SchemaRenameNameConflictError(SchemaTransformError):
+    """Raised when renaming types or fields cause name conflicts.
+
+    This may be raised if a field or type is renamed to conflict with another field or type.
+    """
+
+    name_conflicts: Dict[str, Set[str]]
+
+    def __init__(self, name_conflicts: Dict[str, Set[str]]) -> None:
+        """Record all renaming conflicts."""
+        super().__init__()
+        self.name_conflicts = name_conflicts
+
+    def __str__(self) -> str:
```

Comment:
So far all the exceptions I've seen get initialized with `ExceptionClassName(string_message_goes_here)`. When I searched it up, I found a number of different ways to modify the class (such as modifying `__str__` and/or `__repr__`) so that we get both a human-readable error message and the argument and it seems like this is the best way to do it.

Reply:
The code coverage alert about `utils.py` is also because the `__str__` function doesn't get explicitly called in the tests.

Reply:
The code coverage alert is correct. Consider adding a test for this `__str__` since it isn't trivial.

For example, test coverage would probably have caught the fact that the function currently contains a subtle bug: the two error messages get concatenated without a space between them, resulting in a message like `[...] the resulting schema.Applying the renaming [...]`.

Reply:
Fixed the bug, although I'm not sure on how exactly to test `__str__`. We could look for an exact match for the hardcoded parts (i.e. ignore any non-deterministic issues when converting dicts and sets to strings), but that'd change as soon as the error message changes. At that point it seems like it'd be basically the same as testing that the code raises a `SchemaRenameNameConflictError` at all. 

Alternatively, we could test to see if attributes like `name_conflicts` show up in the error message (which wouldn't catch the bug you mentioned above, but would check that we're reporting the errors correctly). That could also get [messy](https://docs.python.org/3/library/ast.html#ast.literal_eval) especially if there are a lot of errors involved.

Reply:
Dicts are ordered (perhaps sets too?), so I don't think there should be non-determinism issues. I think it's fine to require the test to change if the error message changes.

Here's the underlying thing worth mentioning. Not writing a test for something is implicitly saying "this code isn't buggy, and the code itself is so obvious that debugging a broken test will be harder than just spotting the bug in the code itself." In this case, there was a bug due to some non-trivial logic, and the bug wasn't obvious but would have been easily caught by tests that exercise the various branches. So I think tests are warranted.

Reply:
> Dicts are ordered (perhaps sets too?)

Did some digging and [dicts are ordered](https://stackoverflow.com/questions/39980323/are-dictionaries-ordered-in-python-3-6) according to the spec for Python 3.7+ and as an implementation detail in the CPython implementation for 3.6, but not explicitly part of the spec. I'll implement the tests, but will this potentially cause issues later down the road?

Reply:
Good catch. The question then becomes "are we more likely to support a non-CPython 3.6 version, or drop Python 3.6 first?"

I'd argue the answer is that we're more likely to drop Python 3.6 first.

In general, I don't love depending on implementation details like this, but in this case it seems like the cure would be worse than the disease. I'd argue the prudent choice here would be to get to "good" now vs getting to "perfect" later.

Reply:
Hmm, after some further experimentation and thinking I think this might've been a moot point because insertion order for the dicts and sets ([sets are still unordered anyways](https://stackoverflow.com/questions/45581901/are-sets-ordered-like-dicts-in-python3-6)) depends on the order in which the `GraphQL` visitor visits nodes at the same depth in the schema AST (e.g. one object type vs. the other), and I don't see any guarantees about that in the docs.

~~As a workaround, I'll split the string up and check the parts separately.~~ I'll split the two cases into two separate classes because my initial attempt at writing adequate tests for `__str__` when both cases are covered by `SchemaRenameNameConflictError` ended up requiring a lot of [stringly-typed](https://blog.codinghorror.com/new-programming-jargon/) silliness.